tests/python/relay/aot/test_pass_aot_lower_main.py

```diff
@@ -0,0 +1,435 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import tvm
+import tvm.testing
+from tvm.script import tir as T
+from tvm.relay.backend.aot import AOTLowerMain, CallType
+
+import numpy as np
+import pytest
+
+
+def make_const(dtype, shape):
+    return tvm.relay.const(np.zeros(shape).astype(dtype))
+
+
+def make_consts(dtype, shapes):
+    return [make_const(dtype, shape) for shape in shapes]
+
+
+def transform(mod):
+    host_target = tvm.target.Target("llvm")
+    prim_target = tvm.target.Target("llvm", host=host_target)
+    ctxt = tvm.transform.PassContext()
+    config = tvm.target.make_compilation_config(ctxt, prim_target)
+    mod = tvm.relay.transform.PlanDevices(config)(mod)
+    mod = tvm.relay.transform.InferType()(mod)
+    if mod.get_attr("device_contexts") is None:
```

Comment:
The idea with this is that if you add it in a test it won't be overwritten, but if you don't add it then it'll get set to a sensible default. I agree hiding this logic in a helper func reduces test clarity though, so will look into a more explicit function (or see below...)

side note: I really hate these magic attributes in the compiler and they make tests unnecessarily hard to follow by essentially 'cheating' the IRModule->IRModule property. I'm going to take another look at the refactored pass to see if either I can make `device_contexts` an explicit argument to initialize the pass, or derive it from the IRModule directly.

Reply:
The second option is clearer but the information will still need to be threaded from `LowerTE` to `LowerMain` due to the first pass producing a new `IRModule` without any way of providing the device information (which makes sense given the `IRModule` is meant to contain all this information). Maybe you could also re-scan the `main` function within AOT to re-create the device contexts rather than using attributes?

Reply:
I found your TODO for this haha https://github.com/apache/tvm/blob/a0fe74b3c3608929b21faeaea422ac09aa2f75eb/src/relay/backend/te_compiler.cc#L1230-L1231

It looks like some destructive lowering takes place in `LowerTE` removing `kCompiler` attrs that are otherwise saved into this data structure. One option would be to have a 'device_context' attr defined for each PrimFunc (which I consider a lesser sin that IRModule level attrs).

Reply:
Oh no 🙀  I'm now curious why past-@Mousius didn't attach it as `kTarget` to the `PrimFunc` which leads me to think that there was some other logic that broke due to that 🤔 I think for now attaching a device context instead will be less dangerous.

Reply:
Also, back to my initial question, doesn't this mean that an empty `device_contexts` attr is totally valid for any `IRModule` without a `kCompiler` annotated function which also has a `Target` with `use_device_api` ?